lib/Wikia/tests/Util/SamplerProxyTest.php

```diff
@@ -96,6 +96,115 @@ function testUnsampledMethod() {
 		$this->assertEquals( $testResult, $result );
 	}
 
+
+	function testSampling100Percent() {
+		$this->doSamplingTest( 100, false, 0.0 );
+	}
+
+	function testSampling75Percent() {
+		$this->doSamplingTest( 75, false, 0.1 );
+	}
+
+	function testSampling50Percent() {
+		$this->doSamplingTest( 50, false, 0.1 );
+	}
+
+	function testSampling25Percent() {
+		$this->doSamplingTest( 25, false, 0.1 );
+	}
+
+	function testSampling0Percent() {
+		$this->doSamplingTest( 0, false, 0 );
+	}
+
+	function testShadowing100Percent() {
+		$this->doSamplingTest( 100, true, 0.0 );
+	}
+
+	function testShadowing75Percent() {
+		$this->doSamplingTest( 75, true, 0.1 );
+	}
+
+	function testShadowing50Percent() {
+		$this->doSamplingTest( 50, true, 0.1 );
+	}
+
+	function testShadowing25Percent() {
+		$this->doSamplingTest( 25, true, 0.1 );
+	}
+
+	function testShadowing0Percent() {
+		$this->doSamplingTest( 0, true, 0 );
+	}
+
+	function doSamplingTest( $samplingRate, $enableShadowing, $deviation ) {
+
+		$testOriginalCallable = [ $this->originalMock, $this->originalMethodToSample ];
+		$testAlternateCallable = [ $this->alternateMock, $this->alternateMethod ];
+
+		$testArg1 = 1;
+		$testArg2 = 'two';
+		$testArg3 = array( 4, 5, 6 );
+		$originalTestResult = 'original';
+		$alternateTestResult = 'alternate';
+
+		$builder = SamplerProxy::createBuilder();
+		$samplerProxy =
+			$builder->setEnableShadowing( $enableShadowing )
+				->setMethodSamplingRate( $samplingRate )
+				->setOriginalCallable( $testOriginalCallable )
+				->setAlternateCallable( $testAlternateCallable )
+				->build();
+
+		$originalCallableRecorder =
+			( $samplingRate < 100 || $enableShadowing ) ? $this->atLeastOnce() : $this->never();
+		$this->originalMock->expects( $originalCallableRecorder )
+			->method( $this->originalMethodToSample )
+			->with( $testArg1, $testArg2, $testArg3 )
+			->willReturn( $originalTestResult );
+
+		$alternateCallableRecorder = $samplingRate > 0 ? $this->atLeastOnce() : $this->never();
+		$this->alternateMock->expects( $alternateCallableRecorder )
+			->method( $this->alternateMethod )
+			->with( $testArg1, $testArg2, $testArg3 )
+			->willReturn( $alternateTestResult );
+
+		// Seed the random generator so that we have a consistent dataset of psuedo-random
+		// numbers across the sample set.  We're trying to test the sampling code, not the
+		// random number generator, so this is reasonable and prevents intermittent test failures
+		// caused by a psuedo-random datasets with a highly skewed distribution.
+		srand(123456);
```

Comment:
Good catch.  I didn't realize ```srand()``` had fallen out of favor.  Looks like I can leverage namespaces to override ```rand()``` in tests, but that seems a bit unsafe as well.  I'll check to see if we're already using a mock package that can safely mock system functions - I'd rather not add a method to production code just for testing if it can be avoided.